src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2Client.cs

```diff
@@ -0,0 +1,440 @@
+                            return result;
+                        }
+
+                        var error = JsonConvert.DeserializeObject<ErrorResponseJson>(json);
+
+                        switch (error.Error)
+                        {
+                            case OAuth2Constants.DeviceAuthorization.Errors.AuthorizationPending:
+                                // Retry with the current polling interval value
+                                break;
+                            case OAuth2Constants.DeviceAuthorization.Errors.SlowDown:
+                                // We must increase the polling interval by 5 seconds
+                                retryInterval = retryInterval.Add(TimeSpan.FromSeconds(5));
+                                break;
+                            default:
+                                // For all other errors do not retry
+                                throw CreateExceptionFromResponse(json);
+                        }
+                    }
+                }
+                catch (TimeoutException)
+                {
+                    // Back-off exponentially (2 * x = x + x)
+                    retryInterval += retryInterval;
+                }
+
+                // Wait the polling interval before retrying
+                await Task.Delay(retryInterval);
+            }
+        }
+
+        #endregion
+
+        #region Extension Points
+
+        protected virtual bool TryCreateTokenEndpointResult(string json, out OAuth2TokenResult result)
+        {
+            if (TryDeserializeJson(json, out TokenEndpointResponseJson jsonObj))
+            {
+                result = jsonObj.ToResult();
+                return true;
+            }
+
+            result = null;
+            return false;
+        }
+
+        protected virtual bool TryCreateExceptionFromResponse(string json, out OAuth2Exception exception)
+        {
+            if (TryDeserializeJson(json, out ErrorResponseJson obj))
+            {
+                exception = obj.ToException();
+                return true;
+            }
+
+            exception = null;
+            return false;
+        }
+
+        #endregion
+
+        #region Helpers
+
+        private HttpRequestMessage CreateRequestMessage(HttpMethod method, Uri requestUri, HttpContent content = null, bool addAuthHeader = false)
+        {
+            var request = new HttpRequestMessage(method, requestUri) {Content = content};
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(Constants.Http.MimeTypeJson));
+
+            if (addAuthHeader && !string.IsNullOrEmpty(_clientSecret))
+            {
+                request.AddBasicAuthenticationHeader(_clientId, _clientSecret);
+            }
+
+            return request;
+        }
+
+        private static string GenerateRandomString()
+        {
+            return Guid.NewGuid().ToString("N");
+        }
+
+        private Exception CreateExceptionFromResponse(string json)

```

Comment:
Naming nit: this doesn't just create the exception, it also throws it. Let's rename or have it return the exception for the caller to throw. Latter might be preferable to avoid the extra stack frame?

Reply:
Ah! My bad.. this _should_ be `return`ing the `Exception` here..